File: SQL/Select/SQLSelectTableJoin.vb

' ___________________________________________________
'
'  © Hi-Integrity Systems 2010. All rights reserved.
'  www.hisystems.com.au - Toby Wicks
' ___________________________________________________
'

Option Strict On
Option Explicit On

Namespace SQL

    Public Class SQLSelectTableJoin
        Inherits SQLSelectTableBase

        Public Enum Type
            Inner
            FullOuter
            LeftOuter
            RightOuter
        End Enum

        Private pobjParent As SQLSelectTableJoins
        Private pobjLeftTable As SQLSelectTableBase
        Private pobjRightTable As SQLSelectTableBase
        Private pobjConditions As SQLSelectTableJoinConditions
        Private peType As SQLSelectTableJoin.Type

        Friend Sub New(ByVal objParent As SQLSelectTableJoins)

            MyBase.New()
            pobjParent = objParent
            pobjConditions = New SQLSelectTableJoinConditions(Me)

        End Sub

        Property TheType() As SQLSelectTableJoin.Type
            Get

                Return peType

            End Get

            Set(ByVal Value As SQLSelectTableJoin.Type)

                peType = Value

            End Set
        End Property

        Public Property LeftTable() As SQLSelectTableBase
            Get
                LeftTable = pobjLeftTable
            End Get

            Set(ByVal Value As SQLSelectTableBase)

                If Value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                pobjLeftTable = Value

            End Set
        End Property

        Public Property RightTable() As SQLSelectTableBase
            Get
                Return pobjRightTable
            End Get

            Set(ByVal Value As SQLSelectTableBase)

                If Value Is Nothing Then
                    Throw New ArgumentNullException
                End If

                pobjRightTable = Value

            End Set
        End Property

        Public Property Where() As SQLSelectTableJoinConditions
            Get

                Return pobjConditions

            End Get

            Set(ByVal Value As SQLSelectTableJoinConditions)

                pobjConditions = Value

            End Set
        End Property

        Protected Overrides ReadOnly Property Source(ByVal eConnectionType As Database.ConnectionType) As String
            Get

                Dim strSQL As String = _
                    pobjLeftTable.SQL(eConnectionType) & " " & GetJoinString(Me.TheType) & " " & pobjRightTable.SQL(eConnectionType)

                If Not pobjConditions Is Nothing AndAlso Not pobjConditions.IsEmpty Then
                    strSQL &= " ON " & pobjConditions.SQL(eConnectionType)
                End If

                'Surround the join with parentheses - MS Access won't accept it otherwise
                Return "(" & strSQL & ")"

            End Get
        End Property

        Private Shared Function GetJoinString(eJoinType As Type) As String

            Select Case eJoinType
                Case SQLSelectTableJoin.Type.Inner
                    Return "INNER JOIN"
                Case SQLSelectTableJoin.Type.FullOuter
                    Return "FULL OUTER JOIN"
                Case SQLSelectTableJoin.Type.LeftOuter
                    Return "LEFT OUTER JOIN"
                Case SQLSelectTableJoin.Type.RightOuter
                    Return "RIGHT OUTER JOIN"
                Case Else
                    Throw New NotImplementedException(eJoinType.ToString)
            End Select

        End Function

    End Class

End Namespace

File: SQL/Select/SQLSelectTableJoin.cs
// _________________________________________________________________________
//
//  © Hi-Integrity Systems 2010. All rights reserved.
//  www.hisystems.com.au - Toby Wicks
// 
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//	    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// _________________________________________________________________________
//

using System.Collections;
using System;
using System.Data;

namespace DatabaseObjects.SQL
{
	public class SQLSelectTableJoin : SQLSelectTableBase
	{
		public enum Type
		{
			Inner,
			FullOuter,
			LeftOuter,
			RightOuter
		}
			
		private SQLSelectTableBase pobjLeftTable;
		private SQLSelectTableBase pobjRightTable;
		private SQLSelectTableJoinConditions pobjConditions;
		private SQLSelectTableJoin.Type peType;
			
		internal SQLSelectTableJoin()
		{
			pobjConditions = new SQLSelectTableJoinConditions(this);
		}
			
		public SQLSelectTableJoin.Type TheType
		{
			get
			{
				return peType;
			}
				
			set
			{
				peType = value;
			}
		}
			
		public SQLSelectTableBase LeftTable
		{
			get
			{
                return pobjLeftTable;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				pobjLeftTable = value;
			}
		}
			
		public SQLSelectTableBase RightTable
		{
			get
			{
				return pobjRightTable;
			}
				
			set
			{
				if (value == null)
					throw new ArgumentNullException();
					
				pobjRightTable = value;
			}
		}
			
		public SQLSelectTableJoinConditions Where
		{
			get
			{
				return pobjConditions;
			}
				
			set
			{
				pobjConditions = value;
			}
		}

		internal override string Source(Serializers.Serializer serializer)
		{
			return serializer.SerializeSelectTableJoin(this);
		}
	}
}
